include/bluetooth/services/ddfs.h

```diff
@@ -0,0 +1,266 @@
+ */
+
+#ifndef BT_DDFS_H_
+#define BT_DDFS_H_
+
+#include <stdlib.h>
+#include <zephyr/types.h>
+#include <bluetooth/conn.h>
+#include <bluetooth/uuid.h>
+#include <bluetooth/gatt.h>
+
+/**@file
+ * @defgroup bt_ddfs Direction and Distance Finding Service API
+ * @{
+ * @brief API for the Direction and Distance Finding Service (DDFS).
+ */
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/** @brief UUID of the DDF Service. **/
+#define BT_UUID_DDFS_VAL \
+	BT_UUID_128_ENCODE(0x21490000, 0x494a, 0x4573, 0x98af, 0xf126af76f490)
+
+/** @brief UUID of the Distance Measurement Characteristic. **/
+#define BT_UUID_DDFS_DISTANCE_MEAS_VAL \
+	BT_UUID_128_ENCODE(0x21490001, 0x494a, 0x4573, 0x98af, 0xf126af76f490)
+
+/** @brief UUID of the Azimuth Measurement Characteristic. **/
+#define BT_UUID_DDFS_AZIMUTH_MEAS_VAL \
+	BT_UUID_128_ENCODE(0x21490002, 0x494a, 0x4573, 0x98af, 0xf126af76f490)
+
+/** @brief UUID of the Elevation Measurement Characteristic. **/
+#define BT_UUID_DDFS_ELEVATION_MEAS_VAL \
+	BT_UUID_128_ENCODE(0x21490003, 0x494a, 0x4573, 0x98af, 0xf126af76f490)
+
+/** @brief UUID of the Feature Characteristic. **/
+#define BT_UUID_DDFS_FEATURE_VAL \
+	BT_UUID_128_ENCODE(0x21490004, 0x494a, 0x4573, 0x98af, 0xf126af76f490)
+
+/** @brief UUID of the Control Point Characteristic. **/
+#define BT_UUID_DDFS_CTRL_POINT_VAL \
+	BT_UUID_128_ENCODE(0x21490005, 0x494a, 0x4573, 0x98af, 0xf126af76f490)
+
+#define BT_UUID_DDFS                 BT_UUID_DECLARE_128(BT_UUID_DDFS_VAL)
+#define BT_UUID_DDFS_DISTANCE_MEAS   BT_UUID_DECLARE_128(BT_UUID_DDFS_DISTANCE_MEAS_VAL)
+#define BT_UUID_DFFS_AZIMUTH_MEAS    BT_UUID_DECLARE_128(BT_UUID_DDFS_AZIMUTH_MEAS_VAL)
+#define BT_UUID_DFFS_ELEVATION_MEAS  BT_UUID_DECLARE_128(BT_UUID_DDFS_ELEVATION_MEAS_VAL)
+#define BT_UUID_DDFS_FEATURE         BT_UUID_DECLARE_128(BT_UUID_DDFS_FEATURE_VAL)
+#define BT_UUID_DFFS_CTRL_POINT      BT_UUID_DECLARE_128(BT_UUID_DDFS_CTRL_POINT_VAL)
+
+/** Ranging mode */
+enum bt_ddfs_dm_ranging_mode {
+	/* Round trip timing */
+	BT_DDFS_DM_RANGING_MODE_RTT,
+
+	/* Multi-carrier phase difference */
+	BT_DDFS_DM_RANGING_MODE_MCPD,
+};
+
+enum bt_ddfs_quality {
+	/* Good measurement quality */
+	BT_DDFS_QUALITY_OK,
+
+	/* Poor measurement quality */
+	BT_DDFS_QUALITY_POOR,
+
+	/* Measurement not for use */
+	BT_DDFS_QUALITY_DO_NOT_USE,
+
+	/* Measurement quality not specified */
+	BT_DDFS_QUALITY_NONE,
+};
+
+/** Direction and Distance Service notification events. */
+enum bt_ddfs_evt {
+	/** Distance Measurement notification enable. */
+	BT_DDFS_EVT_DM_NOTIFY_ENABLE,
+
+	/** Distance Measurement notification disable. */
+	BT_DDFS_EVT_DM_NOTIFY_DISABLE,
+
+	/** Azimuth Measurement notification enable. */
+	BT_DDFS_EVT_AM_NOTIFY_ENABLE,
+
+	/** Azimuth Measurement notification disable. */
+	BT_DDFS_EVT_AM_NOTIFY_DISABLE,
+
+	/** Elevation Measurement notification enable. */
+	BT_DDFS_EVT_EM_NOTIFY_ENABLE,
+
+	/** Elevation Measurement notification disable. */
+	BT_DDFS_EVT_EM_NOTIFY_DISABLE,
+};
```

Comment:
Please replace event IDs (nRF5 SDK style) with callback pointer event notification (NCS style). Currently, we mix these two styles inside this module.

I suggest changing the _event_handler_ callback to the new callback with the following signature
_notification_config_changed(enum bt_ddfs_char_id, bool enabled)_

The _event_handler_ callback is very general and in my opinion, should be changed. After all, we could also add event IDs for cp_dm_ranging_mode_set and cp_dm_config_read callback to the _event_handler_ and it would be okay. However, this style was used in nRF5 SDK. I think we should stick to the NCS style which uses function pointers as callbacks for various events.

Reply:
We could even go one step further and remove the char_id parameter and have 3 different callbacks:

- _am_notification_config_changed(bool enabled)_
- _dm_notification_config_changed(bool enabled)_
- _em_notification_config_changed(bool enabled)_

Reply:
Good point. I've added three independent callbacks.